### PR TITLE
fix: first_or_create to find_or_create_by

### PR DESCRIPTION
**Pull request by hui00**
The first_or_create method merely checks if a StaticContent object exists or not. On the other hand, find_or_create_by will check if an object exists with specific arguments, and if it doesn't, it will create one using these arguments as well.

### CODE REVIEW

**Comment by donni106** [comment]
What about putting the `data_type` in the arguments to check to have the correct result. It COULD be possible to have a static content with the same name but html type. The content would not be possible for further use.

**Comment by hui00** [comment]
Okay, we can also include the `data_type`

**Review by donni106 — CHANGES_REQUESTED** [review]
Please include the `data_type` as discussed

File: spec/mailers/previews/defect_report_mailer_preview.rb [hunk 1]
```diff
@@ -11,11 +11,9 @@ def notify_for_category
     generic_item.content_blocks = [OpenStruct.new(body: "Test mit mehr Worten", title: "Test")]
     generic_item.external_reference = OpenStruct.new(unique_id: "1234567890")
 
-    StaticContent.first_or_create(
-      name: "defectreport-notify-for-category",
-      data_type: "json",
-      content: "{\"app_name\":\"Test-App\",\"mail_footer\":[\"Zeile 1\", \"\", \"Zeile 2\"]}"
-    )
+    StaticContent.find_or_create_by(name: 'defectreport-notify-for-category', data_type: "json" ) do |s|
```
Comment (donni106) [inline spec/mailers/previews/defect_report_mailer_preview.rb:14]: please do not mix quotes and use double quotes everywhere. the `name: ...` is single quoted in every file changed

**Review by donni106 — APPROVED** [review]

**Comment by marcometz** [comment]
What is this the fix for?
What was the Problem?

**Comment by hui00** [comment]
> What is this the fix for? What was the Problem?

For the email preview, StaticContent.first_or_create is used only for an existing StaticContent. However, if an incorrect StaticContent exists, it will not create the one we need for the email preview.

**Comment by marcometz** [comment]
> > What is this the fix for? What was the Problem?
> 
> For the email preview, StaticContent.first_or_create only check for an existing StaticContent. However, if an incorrect StaticContent exists, it will not create the one we need for the email preview.

In Ruby on Rails, there are two methods that can be used to attempt to find an object and create it if it cannot be found. These methods are first_or_create and find_or_create_by. Here are the differences between the two:

`first_or_create`: This method first attempts to find the first object that matches the given conditions. If it cannot find such an object, it creates a new one. The conditions are passed in a hash. For example:

```
User.where(name: 'John').first_or_create
```

In this example, it tries to find the first user with the name 'John'. If no such user is found, a new user with this name is created.

`find_or_create_by`: This method is similar to first_or_create, but it allows you to specify the conditions directly in the method, rather than passing them in a hash. For example:

```
User.find_or_create_by(name: 'John')
```

In this example, it also tries to find the first user with the name 'John'. If no such user is found, a new user with this name is created.

The main difference between the two methods is how the conditions are specified. With first_or_create, the conditions must be passed in a hash that is given to the where method, while with find_or_create_by, the conditions can be passed directly as arguments to the method.

**In this particular case, there is a significant difference between the two methods in terms of how the content property is assigned.**

With first_or_create, the content property is assigned every time the method is called, regardless of whether a new object is created or an existing object is found. This means that the content field will be overwritten every time this method is called.

With find_or_create_by and the subsequent block, on the other hand, the content property is assigned only when a new object is created. If an existing object is found, the **content property remains unchanged**.

(powered by chatGPT)


So I am unsure if this was the original intention.

```
StaticContent.first_or_create(
      name: "A",
      data_type: "B",
      content: "C"
    )
```

should result in   `StaticContent.where(name: "A",  data_type: "B", content: "C").first_or_create`
which should be the same as `StaticContent.first_or_create(name: "A",  data_type: "B", content: "C")`

which is a **difference** to `StaticContent.where(name: "A",  data_type: "B").first_or_create(content: "C")`

Could you please check this?
I would love to know this for sure.

Thanks a lot!!!

**Comment by hui00** [comment]
>

**Comment by hui00** [comment]
> should result in `StaticContent.where(name: "A", data_type: "B", content: "C").first_or_create` which should be the same as `StaticContent.first_or_create(name: "A", data_type: "B", content: "C")`

they are not identical.
`StaticContent.first_or_create(name: "A", data_type: "B", content: "C")` will return the first instance of StaticContent it finds. If none exists, it will create one with the provided attributes.

On the other hand
`StaticContent.where(name: "A", data_type: "B", content: "C").first_or_create` will specifically search for a StaticContent with the given attributes. If none is found, it will create one.


> which is a **difference** to `StaticContent.where(name: "A", data_type: "B").first_or_create(content: "C")` which i think is similar to `StaticContent.where(name: "A", data_type: "B").find_or_create_by |a| { a.content: "C"}`

StaticContent.where(name: "A", data_type: "B").first_or_create(content: "C")  This method will create a StaticContent with attributes name: "A", data_type: "B", content: "C" only if there isn't already a StaticContent with name: "A" and data_type: "B".


https://guides.rubyonrails.org/active_record_querying.html#find-or-build-a-new-object
> The `find_or_create_by` version also reads better, frankly.
> The `first_or_create` family of methods have been nodoc'ed in favour  of this API.

https://github.com/rails/rails/commit/0096f53b25e68c3fc79429253f816fff4a4ee596

**Pull request merged**